packages/liveblocks-core/src/lib/position.ts

```diff
@@ -1,133 +1,342 @@
+  } else if (lo > hi) {
+    return _between(hi, lo) as Pos;
+  } else {
+    throw new Error("Cannot compute value between two equal positions");
+  }
 }
 
-function makePositionFromCodes(before: number[], after: number[]): number[] {
+/**
+ * Like between(), but guaranteed that lo < hi.
+ */
+function _between(lo: Pos, hi: Pos | ""): Pos {
   let index = 0;
-  const result = [];
 
+  const loLen = lo.length;
+  const hiLen = hi.length;
   while (true) {
-    const beforeDigit: number = before[index] || min;
-    const afterDigit: number = after[index] || max;
-
-    // istanbul ignore if
-    if (beforeDigit > afterDigit) {
-      throw new Error(
-        `Impossible to generate position between ${before} and ${after}`
-      );
-    }
+    const loCode = index < loLen ? lo.charCodeAt(index) : MIN_CODE;
+    const hiCode = index < hiLen ? hi.charCodeAt(index) : MAX_CODE;
 
-    if (beforeDigit === afterDigit) {
-      result.push(beforeDigit);
+    if (loCode === hiCode) {
       index++;
       continue;
     }
 
-    if (afterDigit - beforeDigit === 1) {
-      result.push(beforeDigit);
-      result.push(...makePositionFromCodes(before.slice(index + 1), []));
-      break;
+    // Difference of only 1 means we'll have to settle this in the next digit
+    if (hiCode - loCode === 1) {
+      const prefix = lo.substring(0, index + 1);
+      const suffix = lo.substring(index + 1) as Pos;
+      const nines = ""; // Will get interpreted like .999999…
+      return (prefix + _between(suffix, nines)) as Pos;
+    } else {
+      // Difference of more than 1 means we take the "middle" between these digits
+      return (takeN(lo, index) +
+        String.fromCharCode((hiCode + loCode) >> 1)) as Pos;
     }
-
-    const mid = (afterDigit + beforeDigit) >> 1;
-    result.push(mid);
-    break;
   }
-
-  return result;
 }
 
-export function posCodes(str: string): number[] {
-  const codes: number[] = [];
-  for (let i = 0; i < str.length; i++) {
-    codes.push(str.charCodeAt(i));
-  }
-  return codes;
+function takeN(pos: string, n: number): string {
+  return n < pos.length
+    ? pos.substring(0, n)
+    : pos + ZERO.repeat(n - pos.length);
 }
 
-export function pos(codes: number[]): string {
-  return String.fromCharCode(...codes);
+/**
+ * Okay, this regex may need a little explanation.
+ *
+ * It checks whether a given string is a valid Pos value. There are three
+ * rules:
+ *
+ *   - All characters in the string must be from our alphabet
+ *   - The string must not have any trailing "zeroes" (trailing " ")
+ *   - The string must not be the empty string
+ *
+ * The first range [\x20-\x7E] checks whether all characters are from the
+ * alphabet. Here, \x20 = 32 (min), and \x7E = 126 (max).
+ *
+ * In the second range, we check [\x21-\x7E], which excludes the " " char
+ * (\x20). This ensures there are no trailing zeroes, _and_ ensures the string
+ * is non-empty.
+ *
+ * Using this regex is faster than iterating over the string
```

Comment:
Curious to see a benchmark, especially on positions with 1 or 2 characters 👀

Reply:
You're right, I think the benchmark I used for this when I wrote this was not representative. Since most Pos strings will be super short, the regex indeed does not beat it, and the perf benefit only starts to kick in with strings from length 5 and up. I've replaced this with a manual check again in 80cbc4a2aaa98f2588765584830cd054115b65e6.